File: benchmarking/typescript_bench/bench_read.ts
#!/usr/bin/env -S npx tsx
/**
 * MCAP read benchmark for TypeScript.
 */

import { open, stat } from "fs/promises";

import {
  McapIndexedReader,
  McapStreamReader,
} from "../../typescript/core/src/index.ts";
import { FileHandleReadable } from "../../typescript/nodejs/src/index.ts";
import { loadDecompressHandlers } from "../../typescript/support/src/decompressHandlers.ts";

async function main(): Promise<number> {
  if (process.argv.length < 3 || process.argv.length > 7) {
    process.stderr.write(
      `Usage: ${process.argv[1]} <input_file> [mode] [num_messages] [payload_size] [filter]\n` +
        `  filter: topic | timerange | topic_timerange\n`,
    );
    return 1;
  }

  const filename = process.argv[2]!;
  const mode = process.argv[3] ?? "unknown";
  const numMessagesStr = process.argv[4] ?? "0";
  const payloadSizeStr = process.argv[5] ?? "0";
  const filter = process.argv[6];

  const decompressHandlers = await loadDecompressHandlers();

  let msgCount = 0;

  // Time file open + reader creation + message iteration
  const tStart = process.hrtime.bigint();

  if (filter != null && filter !== "") {
    // Filtered reads: use McapIndexedReader to leverage the chunk index
    const fileHandle = await open(filename, "r");
    const reader = await McapIndexedReader.Initialize({
      readable: new FileHandleReadable(fileHandle),
      decompressHandlers,
    });

    const readArgs: {
      topics?: string[];
      startTime?: bigint;
      endTime?: bigint;
    } = {};
    if (filter === "topic") {
      readArgs.topics = ["/imu"];
    } else if (filter === "timerange") {
      readArgs.startTime = 3000000000n;
      readArgs.endTime = 5000000000n;
    } else if (filter === "topic_timerange") {
      readArgs.topics = ["/lidar"];
      readArgs.startTime = 4000000000n;
      readArgs.endTime = 6000000000n;
    }

    for await (const message of reader.readMessages(readArgs)) {
      if (message.data.length === 0) {
        process.stderr.write("Empty message\n");
      }
      msgCount++;
    }

    await fileHandle.close();
  } else {
    // Unfiltered reads: use McapStreamReader to handle all file types.
    // Stream the file in chunks to keep memory bounded; append() copies
    // into the reader's internal buffer, so the read buffer is reusable.
    const fileHandle = await open(filename, "r");
    const reader = new McapStreamReader({ decompressHandlers });
    const chunk = new Uint8Array(1024 * 1024);
    for (;;) {
      const { bytesRead } = await fileHandle.read(chunk, 0, chunk.length);
      if (bytesRead === 0) {
        break;
      }
      reader.append(chunk.subarray(0, bytesRead));
      for (;;) {
        const record = reader.nextRecord();
        if (record == null) {
          break;
        }
        if (record.type === "Message") {
          if (record.data.length === 0) {
            process.stderr.write("Empty message\n");
          }
          msgCount++;
        }
      }
    }
    await fileHandle.close();
  }

  const tEnd = process.hrtime.bigint();

  const elapsedNs = tEnd - tStart;
  const wallSec = Number(elapsedNs) / 1e9;
  const fileSize = (await stat(filename)).size;

  const peakRssKb = process.resourceUsage().maxRSS;

  // TSV output: op lang mode num_msgs payload_size file_size elapsed_ns wall_sec peak_rss_kb msg_count
  process.stdout.write(
    `read\ttypescript\t${mode}\t${numMessagesStr}\t${payloadSizeStr}\t${fileSize}\t${elapsedNs}\t${wallSec.toFixed(
      6,
    )}\t${peakRssKb}\t${msgCount}\n`,
  );

  return 0;
}

main().then((code) => process.exit(code));

File: benchmarking/rust_bench/src/bin/bench_read.rs
use std::io::Read;
use std::time::Instant;

use mcap::sans_io::indexed_reader::{IndexedReadEvent, IndexedReader, IndexedReaderOptions};
use mcap::sans_io::linear_reader::{LinearReadEvent, LinearReader};

/// ru_maxrss is KB on Linux but bytes on macOS; normalize to KB.
fn peak_rss_kb() -> libc::c_long {
    let mut rusage: libc::rusage = unsafe { std::mem::zeroed() };
    unsafe { libc::getrusage(libc::RUSAGE_SELF, &mut rusage) };
    if cfg!(target_os = "macos") {
        rusage.ru_maxrss / 1024
    } else {
        rusage.ru_maxrss
    }
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 2 || args.len() > 6 {
        eprintln!(
            "Usage: {} <input_file> [mode] [num_messages] [payload_size] [filter]",
            args[0]
        );
        eprintln!("  filter: topic | timerange | topic_timerange");
        std::process::exit(1);
    }

    let filename = &args[1];
    let mode = if args.len() >= 3 { &args[2] } else { "unknown" };
    let num_messages_str = if args.len() >= 4 { &args[3] } else { "0" };
    let payload_size_str = if args.len() >= 5 { &args[4] } else { "0" };
    let filter = if args.len() >= 6 {
        Some(args[5].as_str())
    } else {
        None
    };

    // Timed: file read + message iteration
    let start = Instant::now();

    let mut msg_count: u64 = 0;

    match filter {
        None => {
            // Stream the file through the sans-io LinearReader to keep
            // memory bounded rather than buffering the whole file. The
            // reader requests a few bytes at a time on unchunked files, so
            // buffer the underlying reads.
            let file = std::fs::File::open(filename).expect("Failed to open file");
            let mut file = std::io::BufReader::with_capacity(1 << 20, file);
            let mut reader = LinearReader::new();
            while let Some(event) = reader.next_event() {
                match event.expect("Failed to read event") {
                    LinearReadEvent::ReadRequest(need) => {
                        let written = file.read(reader.insert(need)).expect("Failed to read file");
                        reader.notify_read(written);
                    }
                    LinearReadEvent::Record { opcode, data } => {
                        if opcode == mcap::records::op::MESSAGE {
                            let record = mcap::parse_record(opcode, data)
                                .expect("Failed to parse message record");
                            if let mcap::records::Record::Message { data, .. } = record {
                                // Touch data to prevent dead-code elimination
                                if data.is_empty() {
                                    eprintln!("Empty message");
                                }
                                msg_count += 1;
                            }
                        }
                    }
                }
            }
        }
        Some(filter_mode) => {
            // The indexed reader operates on byte slices, so the filtered
            // path buffers the whole file. Filtered results do not feed
            // the memory table.
            let buf = std::fs::read(filename).expect("Failed to read file");
            let summary = mcap::Summary::read(&buf)
                .expect("Failed to read summary")
                .expect("No summary found in file");

            let options = match filter_mode {
                "topic" => IndexedReaderOptions::new().include_topics(vec!["/imu"]),
                "timerange" => IndexedReaderOptions::new()
                    .log_time_on_or_after(3_000_000_000)
                    .log_time_before(5_000_000_000),
                "topic_timerange" => IndexedReaderOptions::new()
                    .include_topics(vec!["/lidar"])
                    .log_time_on_or_after(4_000_000_000)
                    .log_time_before(6_000_000_000),
                other => {
                    eprintln!("Unknown filter mode: {}", other);
                    std::process::exit(1);
                }
            };

            let mut reader = IndexedReader::new_with_options(&summary, options)
                .expect("Failed to create indexed reader");
            while let Some(event) = reader.next_event() {
                match event.expect("Failed to read event") {
                    IndexedReadEvent::ReadChunkRequest { offset, length } => {
                        let chunk_data = &buf[offset as usize..][..length];
                        reader
                            .insert_chunk_record_data(offset, chunk_data)
                            .expect("Failed to insert chunk data");
                    }
                    IndexedReadEvent::Message { header: _, data } => {
                        msg_count += 1;
                        if data.is_empty() {
                            eprintln!("Empty message");
                        }
                    }
                }
            }
        }
    }

    let elapsed = start.elapsed();
    let elapsed_ns = elapsed.as_nanos();
    let wall_sec = elapsed.as_secs_f64();

    let file_size = std::fs::metadata(filename)
        .expect("Failed to stat file")
        .len();

    let peak_rss_kb = peak_rss_kb();

    // TSV output: op lang mode num_msgs payload_size file_size elapsed_ns wall_sec peak_rss_kb
    // msg_count
    println!(
        "read\trust\t{}\t{}\t{}\t{}\t{}\t{:.6}\t{}\t{}",
        mode, num_messages_str, payload_size_str, file_size, elapsed_ns, wall_sec, peak_rss_kb,
        msg_count
    );
}
